service/controller/clusterapi/v27/resource/ipam/create.go

```diff
@@ -53,57 +52,37 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 	r.logger.LogCtx(ctx, "level", "debug", "message", "finding out if subnet needs to be allocated for cluster")
 
 	if legacykey.StatusNetworkCIDR(cr) == "" {
-		var statusAZs []v1alpha1.AWSConfigStatusAWSAvailabilityZone
 		var subnetCIDR net.IPNet
 		{
 			r.logger.LogCtx(ctx, "level", "debug", "message", "allocating cluster subnet CIDR")
 
-			subnetCIDR, err = r.allocateSubnet(ctx)
-			if err != nil {
-				return microerror.Mask(err)
-			}
-
 			randomAZs, err := r.selectRandomAZs(legacykey.SpecAvailabilityZones(cr))
 			if err != nil {
 				return microerror.Mask(err)
 			}
 
-			statusAZs, err = splitSubnetToStatusAZs(subnetCIDR, randomAZs)
-			if err != nil {
-				return microerror.Mask(err)
+			callbacks := network.AllocationCallbacks{
+				GetReservedNetworks: r.getReservedSubnets,
+				PersistAllocatedNetwork: func(ctx context.Context, subnet net.IPNet) error {
```

Comment:
It would make sense to me to have something like this. 

```go
callbacks := network.AllocationCallbacks{
  GetReservedNetworks:     r.getReservedNetworks,
  PersistAllocatedNetwork: r.persistAllocatedNetwork,
}
```

Reply:
That would require saving current CR object and selected AZs either in resource instance or context and to me that looks very dirty. I can create a function like this:

```go
func (r *Resource) persistAllocatedNetwork(ctx context.Context, cr v1alpha1.AWSConfig, subnet net.IPNet, azs []string) (func(context.Context, net.IPNet) error) {
    return func(ctx context.Context, subnet net.IPNet) error {
        return r.splitAndPersistReservedSubnet(ctx, cr, subnet, randomAZs)
    }
}
```

So then this could be:

```go
callbacks := network.AllocationCallbacks{
  GetReservedNetworks:     r.getReservedNetworks,
  PersistAllocatedNetwork: r.persistAllocatedNetwork(ctx, cr, subnet, randomAZs),
}
```

But I think that would be more complicated to read and I think on first glance it would probably mislead the reader.

Reply:
My intention was to suggest some more consistent approach for both callbacks where naming and call/configuration is similar. I do not want to be too picky. What you suggested in your last comment looks good, even though my OCD would like to see the same invocation for both callbacks.